File: merge-sort.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"
)

func main() {
	arrayToSort := getArrayToSort()

	sorted := MergeSort(arrayToSort)

	fmt.Println(sorted[:10])
}

func getArrayToSort() []int {
	arry := make([]int, 250001)

	resource := "/Users/nowak/Development/go/src/github.com/oliver-nowak/nos/large_avg.txt" // 2015/10/04 11:47:42 selectionsort took 34.407796662s

	f, err := os.Open(resource)
	if err != nil {
		log.Fatal(err)
	}
	r := bufio.NewReader(f)

	scanner := bufio.NewScanner(r)
	idx := 0

	// read each line
	for scanner.Scan() {
		txt := scanner.Text()
		i, _ := strconv.Atoi(txt)
		arry[idx] = i
		idx += 1
	}

	return arry
}

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}

func MergeSort(A []int) []int {
	var arrayLength = len(A)

	// Base case
	if arrayLength == 1 {
		return A
	}

	// Split the arrays in half
	var left = make([]int, arrayLength/2)
	var right = make([]int, arrayLength/2+1)

	// Recurse to split in half
	left = MergeSort(left)
	right = MergeSort(right)

	// After the recursive split has completed and returned, merge the two arrays back up the stack
	return Merge(left, right)
}

func Merge(left []int, right []int) []int {

	var sorted []int

	for len(left) > 0 && len(right) > 0 {

	}

	return sorted
}

File: merge-sort.py
from numpy import loadtxt


def merge(left, right):
    sorted = []

    # iterate through left and right arrays
    # and compare elements
    while len(left) > 0 and len(right) > 0:

        if left[0] <= right[0]:
            i = left[0]
            sorted.append(i)

            left = left[1:]
        else:
            j = right[0]
            sorted.append(j)

            right = right[1:]

    # iterate through whats left;
    # generally only one of these while statements is executed
    while len(left) > 0:
        i = left[0]
        sorted.append(i)

        left = left[1:]

    while len(right) > 0:
        j = right[0]
        sorted.append(j)

        right = right[1:]

    return sorted



def mergesort(arrayToSort):
    arrayLength = len(arrayToSort)
    middle = arrayLength / 2

    # Base case
    if arrayLength == 1:
        return arrayToSort

    # Split array to sort in half
    left = arrayToSort[0:middle]
    right = arrayToSort[middle:arrayLength]

    # Recurse to continue splitting in half until we hit the base case
    left = mergesort(left)
    right = mergesort(right)

    # Merge the split arrays
    return merge(left, right)


arrayToSort = loadtxt("/Users/nowak/Development/go/src/github.com/oliver-nowak/nos/small_worst.txt", dtype='int')

print arrayToSort

arraySorted = mergesort(arrayToSort)

print arraySorted

